src/tests/ftest/tags.py

```diff
@@ -0,0 +1,527 @@
+        paths (list, optional): path(s) to get tags for. Defaults to all ftest python files
+    """
+    if not paths:
+        paths = all_python_files(FTEST_DIR)
+    for file_path, classes in iter(FtestTagMap(paths)):
+        short_file_path = re.findall(r'ftest/(.*$)', file_path)[0]
+        print(f'{short_file_path}:')
+        for class_name, functions in classes.items():
+            print(f'  {class_name}:')
+            all_methods = []
+            longest_method_name = 0
+            for method_name, tags in functions.items():
+                longest_method_name = max(longest_method_name, len(method_name))
+                all_methods.append((method_name, tags))
+            for method_name, tags in all_methods:
+                method_name_fm = method_name.ljust(longest_method_name, " ")
+                tags_fm = ",".join(sorted_tags(tags))
+                print(f'    {method_name_fm} - {tags_fm}')
+
+
+def files_to_tags(paths):
+    """Get the unique tags for paths.
+
+    Args:
+        paths (list): paths to get tags for
+
+    Returns:
+        set: set of test tags representing paths
+    """
+    # Get tags for ftest paths
+    ftest_tag_map = FtestTagMap(all_python_files(FTEST_DIR))
+    ftest_tag_set = ftest_tag_map.minimal_tags(paths)
+
+    # Future work will also get recommended tags for non-ftest files
+    return ftest_tag_set
+
+
+def run_list(paths):
+    """List unique tags for paths.
+
+    Args:
+        paths (list): paths to list tags of
+    """
+    tags = files_to_tags(paths)
+    print(' '.join(sorted(tags)))
+
+
+def test_tags_util(verbose=False):
+    """Run unit tests for FtestTagMap.
+
+    Can be ran directly as:
+        tags.py unit
+    Or with pytest as:
+        python3 -m pytest tags.py
+
+    Args:
+        verbose (bool): whether to print verbose output for debugging
+    """
+    # pylint: disable=protected-access
+    print('Ftest Tags Utility Unit Tests')
+    tag_map = FtestTagMap([])
+    os.chdir('/')
+
+    def print_step(*args):
+        """Print a step."""
+        print('  ', *args)
+
+    def print_verbose(*args):
+        """Conditionally print if verbose is True."""
+        if verbose:
+            print('    ', *args)
+
+    print_step('__norm_path')
+    assert tag_map._FtestTagMap__norm_path('foo') == '/foo'
+    assert tag_map._FtestTagMap__norm_path('/foo') == '/foo'
+
+    print_step('__parse_avocado_tags')
+    assert tag_map._FtestTagMap__parse_avocado_tags('') == set()
+    assert tag_map._FtestTagMap__parse_avocado_tags('not tags') == set()
+    assert tag_map._FtestTagMap__parse_avocado_tags('avocado tags=foo') == set()
+    assert tag_map._FtestTagMap__parse_avocado_tags(':avocado: tags=foo') == set(['foo'])
+    assert tag_map._FtestTagMap__parse_avocado_tags(':avocado: tags=foo,bar') == set(['foo', 'bar'])
+    assert tag_map._FtestTagMap__parse_avocado_tags(
+        ':avocado: tags=foo,bar\n:avocado: tags=foo2,bar2') == set(['foo', 'bar', 'foo2', 'bar2'])
```

Comment:
Should we test some abnormal cases here too?

- duplicate tags
- avocado tag definitions ending in `,`
- empty avocado tag definitions
- etc.

What about the normal case of a full test method function definition (what is processed on L213)?

Reply:
> duplicate tags

Yeah, we should. Since they are internally stored in a set, duplicates should be handled automatically.
That also makes me think that for linting we might want to check... But that is more difficult because if the internal mapping silently squashes duplicates we don't know.

> avocado tag definitions ending in ,

Yes! I forgot about that one. The linter has a check for this by looking for an "empty" tag. But ideally a definition ending with `,` shouldn't produce an empty tag. I need to update the core logic for this, but left the lint check in for now as the only safeguard.

> empty avocado tag definitions

Yes again. For unit testing / core logic, maybe it should be silently skipped? Or maybe it should raise an exception? For linting, I think an error.

> the normal case of a full test method function definition

+1. Once this PR gets through I'll add a bunch more tests. We might even want to put them in a separate file if they continue to grow.